operator/controllers/remote.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	remoteCluster := &custom.ClusterClient{DefaultClient: nativeClient}
+	restConfig, err := remoteCluster.GetRestConfig(ctx, kymaOwner, nativeObject.Namespace)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	customClient, err := remoteCluster.GetNewClient(restConfig, client.Options{
+		Scheme: nativeClient.Scheme(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	remoteInterface := &RemoteInterface{
+		NativeObject: nativeObject,
+		NativeClient: nativeClient,
+		CustomClient: customClient,
+	}
+
+	if meta.IsNoMatchError(err) {
+		if err := remoteInterface.CreateCRD(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check for remote resource
+	remoteManifest, err := remoteInterface.GetLatestRemote(ctx)
+
+	if client.IgnoreNotFound(err) != nil {
+		// unknown error
+		return nil, nil, err
+	} else if err != nil {
+		// remote resource doesn't exist
+		remoteManifest, err = remoteInterface.createRemote(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check finalizer
+	if !controllerutil.ContainsFinalizer(remoteManifest, labels.ManifestFinalizer) {
+		controllerutil.AddFinalizer(remoteManifest, labels.ManifestFinalizer)
+		if err = remoteInterface.Update(ctx, remoteManifest); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return remoteInterface, remoteManifest, nil
+}
+
+func (r *RemoteInterface) createRemote(ctx context.Context) (*v1alpha1.Manifest, error) {
+	remoteManifest := &v1alpha1.Manifest{}
+	remoteManifest.Name = r.NativeObject.Name
+	remoteManifest.Namespace = r.NativeObject.Namespace
+	remoteManifest.Spec = *r.NativeObject.Spec.DeepCopy()
+	controllerutil.AddFinalizer(remoteManifest, labels.ManifestFinalizer)
+	err := r.CustomClient.Create(ctx, remoteManifest)
+	return remoteManifest, err
+}
+
+func (r *RemoteInterface) SpecSyncToNative(ctx context.Context, remoteManifest *v1alpha1.Manifest) (bool, error) {
+	if remoteManifest.Status.ObservedGeneration == 0 {
+		// observed generation missing - remote resource freshly installed!
+		return true, nil
+	}
+
+	if remoteManifest.Status.ObservedGeneration != remoteManifest.Generation {
+		// outdated
+		r.NativeObject.Spec = *remoteManifest.Spec.DeepCopy()
+		remoteManifest.Status.ObservedGeneration = remoteManifest.Generation
+		return false, r.UpdateStatus(ctx, remoteManifest)
+	}
+
+	return true, nil
+}
+
+func (r *RemoteInterface) StateSyncToRemote(ctx context.Context, remoteManifest *v1alpha1.Manifest) error {
+	if r.NativeObject.Status.State != remoteManifest.Status.State {
+		remoteManifest.Status.State = r.NativeObject.Status.State
+		remoteManifest.Status.ObservedGeneration = remoteManifest.Generation
+		return r.UpdateStatus(ctx, remoteManifest)
+	}
+	return nil
+}
+
+func (r *RemoteInterface) SpecSyncToRemote(ctx context.Context, remoteManifest *v1alpha1.Manifest) (*v1alpha1.Manifest, error) {
+	// ignore cases where native object has never been reconciled!
+	if r.NativeObject.Status.ObservedGeneration == 0 {
```

Comment:
I think this case is also not strictly necessary, since even if it was not reconciled you dont technically need to avoid the next path. If the object in KCP has ObservedGeneration 0 then it will NEVER match with the Native Object Generation anyways

Reply:
Yes not strictly necessary but it also doesn't make sense to proceed on a non-reconciled remote object IMO

Reply:
But dont you check exactly for this in the second if statement? That would be false then right?